resolver/resolver-tcp.go

```diff
@@ -0,0 +1,359 @@
+package resolver
+
+import (
+	"context"
+	"crypto/tls"
+	"net"
+	"sync/atomic"
+	"time"
+
+	"github.com/miekg/dns"
+	"github.com/safing/portbase/log"
+	"github.com/safing/portmaster/netenv"
+	"github.com/tevino/abool"
+)
+
+const (
+	tcpWriteTimeout    = 1 * time.Second
+	ignoreQueriesAfter = 10 * time.Minute
+)
+
+// TCPResolver is a resolver using just a single tcp connection with pipelining.
+type TCPResolver struct {
+	BasicResolverConn
+
+	clientTTL      time.Duration
+	dnsClient      *dns.Client
+	dnsConnection  *dns.Conn
+	connInstanceID *uint32
+
+	queries         chan *dns.Msg
+	inFlightQueries map[uint16]*InFlightQuery
+	clientStarted   *abool.AtomicBool
+}
+
+// InFlightQuery represents an in flight query of a TCPResolver.
+type InFlightQuery struct {
+	Query          *Query
+	Msg            *dns.Msg
+	Response       chan *dns.Msg
+	Resolver       *Resolver
+	Started        time.Time
+	ConnInstanceID uint32
+}
+
+// MakeCacheRecord creates an RCache record from a reply.
+func (ifq *InFlightQuery) MakeCacheRecord(reply *dns.Msg) *RRCache {
+	return &RRCache{
+		Domain:      ifq.Query.FQDN,
+		Question:    ifq.Query.QType,
+		Answer:      reply.Answer,
+		Ns:          reply.Ns,
+		Extra:       reply.Extra,
+		Server:      ifq.Resolver.Server,
+		ServerScope: ifq.Resolver.ServerIPScope,
+	}
+}
+
+// NewTCPResolver returns a new TPCResolver.
+func NewTCPResolver(resolver *Resolver) *TCPResolver {
+	var instanceID uint32
+	return &TCPResolver{
+		BasicResolverConn: BasicResolverConn{
+			resolver: resolver,
+		},
+		clientTTL: defaultClientTTL,
+		dnsClient: &dns.Client{
+			Net:          "tcp",
+			Timeout:      defaultConnectTimeout,
+			WriteTimeout: tcpWriteTimeout,
+		},
+		connInstanceID:  &instanceID,
+		queries:         make(chan *dns.Msg, 100),
+		inFlightQueries: make(map[uint16]*InFlightQuery),
+		clientStarted:   abool.New(),
+	}
+}
+
+// UseTLS enabled TLS for the TCPResolver. TLS settings must be correctly configured in the Resolver.
+func (tr *TCPResolver) UseTLS() *TCPResolver {
+	tr.dnsClient.Net = "tcp-tls"
+	tr.dnsClient.TLSConfig = &tls.Config{
+		MinVersion: tls.VersionTLS12,
+		ServerName: tr.resolver.VerifyDomain,
+		// TODO: use portbase rng
+	}
+	return tr
+}
+
+func (tr *TCPResolver) client(workerCtx context.Context) error { //nolint:gocognit,gocyclo // TODO
```

Comment:
This function is so huge I'm feeling lost while reviewing it. Seems like there are multiple things that could be extracted into dedicated functions (like re-injecting queries).  IMHO this seems over-engineered and is likely not unit-testable (or debug-able :D). To be easily understandable (and thus likely less buggy) a function should not be longer ~25 lines and don't have more than 3-4 levels of indention. Let's split up this one :)